module/os/freebsd/spl/spl_page.c

```diff
@@ -0,0 +1,221 @@
+ * and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+/*
+ * Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
+ * Use is subject to license terms.
+ */
+
+/* Copyright (c) 1984, 1986, 1987, 1988, 1989 AT&T */
+/* All Rights Reserved   */
+
+/*
+ * University Copyright- Copyright (c) 1982, 1986, 1988
+ * The Regents of the University of California
+ * All Rights Reserved
+ *
+ * University Acknowledgment- Portions of this document are derived from
+ * software developed by the University of California, Berkeley, and its
+ * contributors.
+ */
+
+/*
+ * $FreeBSD$
+ */
+
+#include <sys/page.h>
+#include <sys/byteorder.h>
+#include <sys/lock.h>
+#include <sys/vm.h>
+#include <vm/vm_map.h>
+#include <sys/zfs_context.h>
+
+/*
+ *
+ * if the operation is marked as read then we are stating the kernel will
+ * be writing to pages and must give it write access.
+ *
+ *
+ * This code comes from the FreeBSD's linuxkpi and models their implementation
+ * of get_user_pages(), set_page_dirty(), and put_page().
+ */
+long
+zfs_hold_pages(unsigned long start, unsigned long nr_pages, int read,
+    zfs_page_p *pages)
+{
+	vm_map_t map;
+	vm_prot_t prot;
+	size_t len;
+	int count;
+	int write = read ? 1 : 0;
+
+	map = &curthread->td_proc->p_vmspace->vm_map;
+
+	prot = write ? (VM_PROT_READ | VM_PROT_WRITE) : VM_PROT_READ;
+	len = ((size_t)nr_pages) << PAGE_SHIFT;
+	count = vm_fault_quick_hold_pages(map, start, len, prot, pages,
+	    nr_pages);
+
+	if (count == -1)
+		return (-EFAULT);
+	return (count);
+}
+
+long
+zfs_get_user_pages(unsigned long start, unsigned long nr_pages, int read,
+    zfs_page_p *pages)
+{
+	int count;
+
+	count = zfs_hold_pages(start, nr_pages, read, pages);
+
+	if (count == -1)
+		return (-EFAULT);
+
+	ASSERT(count == nr_pages);
+
+	for (int i = 0; i != nr_pages; i++) {
+		zfs_page_p pg = pages[i];
+		vm_page_lock(pg);
+		vm_page_wire(pg);
+		vm_page_unhold(pg);
+		vm_page_unlock(pg);
+	}
+
+	return (nr_pages);
+}
+
+static void
+zfs_set_user_page_dirty(zfs_page_p pp, int read)
+{
+	if (read && pp)
+		vm_page_dirty(pp);
+}
+
+void
+zfs_put_user_pages(zfs_page_p *pages, unsigned long nr_pages,
+    boolean_t read)
+{
+	int count;
+
+	count = zfs_hold_pages((unsigned long)pages[0], nr_pages, read, pages);
+	if (count == -1)
+		return;
+
+	for (int i = 0; i != nr_pages; i++) {
+		zfs_page_p pg = pages[i];
+		vm_page_lock(pg);
+		zfs_set_user_page_dirty(pg, read);
+		vm_page_unwire_noq(pg);
+		vm_page_unhold(pg);
+		vm_page_unlock(pg);
+	}
+}
+
+void
+zfs_set_page_to_stable(zfs_page_p page)
+{
+	ASSERT3P(page, !=, NULL);
```

Comment:
This will cause a panic if it has been exclusively busied by another direct I/O in progress or if the page is shared busy.
```
#define	vm_page_xbusy(m) do {						\
	if (!vm_page_tryxbusy(m))					\
		panic("%s: page %p failed exclusive busying", __func__,	\
		    (m));						\
} while (0)
```

Reply:
You'll notice that nothing in the FreeBSD kernel calls that...

Reply:
You updated this in the current version I just pushed.